_posts/2017-06-15-grpc-load-balancing.md

```diff
@@ -0,0 +1,190 @@
+    <td>Client Side</td>
+  </tr>
+  <tr>
+    <td>Pros</td>
+    <td>
+    
+* Simple client
+* No client-side awareness of backend
+* Works with untrusted clients
+</td>
+    <td>
+    
+* High performance because elimination of extra hop
+</td>
+  </tr>
+  <tr>
+    <td>Cons</td>
+    <td>
+    
+* LB is in the data path
+* Higher latency
+* LB throughput may limit scalability
+</td>
+    <td>
+    
+* Complex client
+* Client keeps track of server load and health
+* Client implements load balancing algorithm
+* Per-language implementation and maintenance burden
+* Client needs to be trusted, or the trust boundary needs to be handled by a lookaside LB.</td>
+  </tr>
+</table>
+
+
+### Proxy Load Balancer options - L3/L4 (Transport) vs L7 (Application) level
+
+Proxy load balancing can be L3/L4 (transport level) or L7 (application level). In transport level load balancing, the server terminates the TCP connection and opens another connection to the backend of choice. The application data (HTTP/2 and gRPC frames) are simply copied between the client connection to the backend connection. L3/L4 LB by design does very little processing, adds less latency compared with  L7 LB and consumes less resources (cheaper). 
+
+In L7 (application level) load balancing, the LB terminates and parses the HTTP/2 protocol. It creates a new HTTP/2 connection to the backend of choice. It then forwards the HTTP/2 streams received from the client to the backend(s) of choice. With HTTP/2 LB can distribute the streams from one client among multiple backends. LB can inspect the request and assign backend based on request contents. As an example, a session cookie sent as part of HTTP header can be used to associate with a specific backend, so all requests for that session are served by the same backend.
+
+#### L3/L4 (Transport) vs L7 (Application)
+
+<table>
+  <tr>
+    <td>
+    Use case
+    </td>
+    <td>
+    Recommendation
+    </td>
+  </tr>
+  <tr>
+    <td>RPC load varies a lot among connections</td>
+    <td>Use Application level LB</td>
+  </tr>
+  <tr>
+    <td>Storage or compute affinity is important</td>
+    <td>Use Application level LB and use cookies or similar for routing requests to correct backend</td>
+  </tr>
+  <tr>
+    <td>Minimizing resource utilization in proxy is more important than features</td>
+    <td>Use L3/L4 LB</td>
+  </tr>
+  <tr>
+    <td>Latency is paramount</td>
+    <td>Use L3/L4 LB</td>
+  </tr>
+</table>
+
+
+### Client side LB options - Thick client vs Look-aside
+
+*Note: look-aside is also known as one-arm in some literature.*
+
+#### Thick client
+
+A thick client approach means the load balancing smarts are implemented in the client. The client is responsible for keeping track of available servers, their workload, and the algorithms used for choosing servers. The client typically integrates libraries that communicate with other infrastructures - Service Discovery, Name Resolution, Quota management, etc.
```

Comment:
with other infrastructures such as service discovery, name resolution, and quota management.